File: python/file_path.py
import pathlib
import errors


class FilePath:
    __create_key = object()

    @classmethod
    def from_absolute(cls, absolute_path: str):
        path = pathlib.PurePosixPath(absolute_path)
        if not path.is_absolute():
            raise errors.ArgumentError("Invalid path, should be absolute")
        parts = path.parts[1:]
        if len(parts) < 2:
            raise errors.ArgumentError("Invalid path, missing name of the file")
        return FilePath(cls.__create_key, parts[0], pathlib.PurePosixPath(*parts))

    @classmethod
    def from_normalized(cls, normalized_path: str):
        path = pathlib.PurePosixPath(normalized_path)

        return FilePath(cls.__create_key, path.parts[0], path)

    def __init__(self, create_key, user_id: str, normalized_path: pathlib.PurePosixPath):
        assert (create_key == FilePath.__create_key), \
            "FilePath objects must be created using FilePath.from_absolute or FilePath.from_normalized functions"
        self._user_id = user_id
        self._normalized = normalized_path

    @property
    def absolute(self):
        return pathlib.PurePosixPath("/").joinpath(self._normalized).as_posix()

    @property
    def normalized(self):
        return self._normalized.as_posix()

    @property
    def user_id(self):
        return self._user_id

    @property
    def basename(self):
        return self._normalized.name

File: typescript/src/filepath.ts
import * as path from "path";
import { ArgumentError } from "./errors";

export class FilePath {
    /**
     * @param pathParts Normalized path split into parts
     * @param normalizedPath Full normalized path
     *
     * @throws ArgumentError when the provided string path is not valid.
     */
    private constructor(pathParts: string[], normalizedPath: string) {
        // At least 2 elements, first empty for the user ID
        // second for the file name
        if (pathParts.length < 2) {
            throw new ArgumentError("Invalid path, missing parts of the path.");
        }
        this.user = pathParts[0];

        // Do NOT add leading slash, S3 does not like it
        this.normPath = normalizedPath;
    }

    public static fromNormalized(normalizedPath: string): FilePath {
        const pathParts = normalizedPath.split(path.posix.sep);
        return new FilePath(pathParts, normalizedPath);
    }

    public static fromAbsolute(absolutePath: string): FilePath {
        if (!path.posix.isAbsolute(absolutePath)) {
            throw new ArgumentError("Invalid path, should be absolute.");
        }

        const normalizedPath = path.posix.normalize(absolutePath);

        const pathParts = normalizedPath.split(path.posix.sep);
        // Remove the leading empty string which was before the first / separator, /userID/...
        pathParts.shift();
        return new FilePath(pathParts, normalizedPath);
    }

    public get absolutePath(): string {
        return `/${this.normPath}`;
    }

    public get normalizedPath(): string {
        return this.normPath;
    }

    public get basename(): string {
        return path.posix.basename(this.normPath);
    }

    public get userID(): string {
        return this.user;
    }

    private user: string;
    private normPath: string;
}
